services/motion/explore/explore.go

```diff
@@ -0,0 +1,574 @@
+	if err != nil {
+		return false, err
+	}
+
+	// Create kinematic base
+	kb, err := ms.createKinematicBase(ctx, componentName, motionCfg)
+	if err != nil {
+		return false, err
+	}
+
+	// Create motionplan plan
+	plan, err := ms.createMotionPlan(ctx, kb, destination, extra)
+	if err != nil {
+		return false, err
+	}
+
+	// Start background processes
+	cancelCtx, cancel := context.WithCancel(ctx)
+	ms.processCancelFunc = cancel
+	defer ms.processCancelFunc()
+
+	// Start polling for obstacles
+	ms.backgroundWorkers.Add(1)
+	goutils.ManagedGo(func() {
+		ms.checkForObstacles(cancelCtx, obstacleDetectors, kb, plan, motionCfg.ObstaclePollingFreqHz)
+	}, ms.backgroundWorkers.Done)
+
+	// Start executing plan
+	ms.backgroundWorkers.Add(1)
+	goutils.ManagedGo(func() {
+		ms.executePlan(cancelCtx, kb, plan)
+	}, ms.backgroundWorkers.Done)
+
+	// this ensures that if the context is cancelled we always return early at the top of the loop
+	if err := ctx.Err(); err != nil {
+		return false, err
+	}
+
+	select {
+	// if context was cancelled by the calling function, error out
+	case <-ctx.Done():
+		return false, ctx.Err()
+
+	// once execution responds: return the result to the caller
+	case resp := <-ms.executionResponseChan:
+		ms.logger.Debugf("execution completed: %s", resp)
+		if resp.err != nil {
+			return resp.success, resp.err
+		}
+		if resp.success {
+			return true, nil
+		}
+
+	// if the checkPartialPlan process hit an error return it, otherwise exit
+	case resp := <-ms.obstacleResponseChan:
+		ms.logger.Debugf("obstacle response: %s", resp)
+		if resp.err != nil {
+			return resp.success, resp.err
+		}
+		if resp.success {
+			return false, nil
+		}
+	}
+	return false, errors.New("no obstacle was seen during movement to goal and goal was not reached")
+}
+
+// checkForObstacles will continuously monitor the generated transient worldState for obstacles in the given
+// motionplan plan. A response will be sent through the channel if an error occurs, the motionplan plan
+// completes or an obstacle is detected in the given range.
+func (ms *explore) checkForObstacles(
+	ctx context.Context,
+	obstacleDetectors []obstacleDetectorObject,
+	kb kinematicbase.KinematicBase,
+	plan motionplan.Plan,
+	obstaclePollingFrequencyHz float64,
+) {
+	// Constantly check for obstacles in path at desired obstacle polling frequency
+	ticker := time.NewTicker(time.Duration(int(1000/obstaclePollingFrequencyHz)) * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			currentPose := spatialmath.NewZeroPose()
+
+			// Look for new transient obstacles and add to worldState
+			worldState, err := ms.generateTransientWorldState(ctx, obstacleDetectors)
+			if err != nil {
+				ms.logger.Debugf("issue occurred generating transient worldState: %v", err)
+			}
+
+			// Check motionplan plan for transient obstacles
+			collisionPose, err := motionplan.CheckPlan(
+				kb.Kinematics(),
+				plan,
```

Comment:
@nfranczak is correct I believe

> hmmmmmmmm, I believe this implies that we could be in collision with an obstacle that is actually behind the robot, right?

If we pass in the entire plan each time, then if the robot drives forwards and someone walks behind it, intersecting the path that the robot has already executed but within `validObstacleDistanceMM`, it will count that as a collision as far as I can tell.

Reply:
Updated plan execution to keep track of the plan step we are at and to only send in the plan from that point onwards